src/server.rs

```diff
@@ -1,213 +1,492 @@
-        let _ = swap.comsig.verify(&swap.onion.commit, &swap.msg)
-            .map_err(|_| jsonrpc_core::Error::invalid_params("ComSignature invalid"))?;
-    
-        let peeled = onion::peel_layer(&swap.onion, &self.server_key)
-            .map_err(|e| jsonrpc_core::Error::invalid_params(e.message()))?;
-        SERVER_STATE.lock().unwrap().push(Submission{
-            excess: peeled.0.excess,
-            input_commit: swap.onion.commit,
-            onion: peeled.1
-        });
-        Ok(Value::String("success".into()))
-    }
-    
+	/// Implements the 'swap' API
+	fn swap(&self, swap: SwapReq) -> Result<Value> {
+		// milestone 3: check that enc_payloads length matches number of configured servers
+		if swap.onion.enc_payloads.len() != 1 {
+			return Err(Error::invalid_params(
+				"Multi server not supported until milestone 3",
+			));
+		}
+
+		// Verify commitment signature to ensure caller owns the output
+		let serialized_onion = swap.onion.serialize().map_err(|e| Error {
+			message: e.to_string(),
+			code: ErrorCode::InternalError,
+			data: None,
+		})?;
+		let _ = swap
+			.comsig
+			.verify(&swap.onion.commit, &serialized_onion)
+			.map_err(|_| Error::invalid_params("ComSignature invalid"))?;
+
+		// Verify that commitment is unspent
+		let input = node::build_input(&self.node, &swap.onion.commit).map_err(|e| Error {
+			message: e.to_string(),
+			code: ErrorCode::InternalError,
+			data: None,
+		})?;
+		let input = input.ok_or(Error::invalid_params("Commitment not found"))?;
+
+		let peeled = swap
+			.onion
+			.peel_layer(&self.server_config.key)
+			.map_err(|e| Error::invalid_params(e.message()))?;
+
+		// Verify the fee meets the minimum
+		let fee: u64 = peeled.0.fee.into();
+		if fee < self.get_minimum_swap_fee() {
+			return Err(Error::invalid_params("Fee does not meet minimum"));
+		}
+
+		// Calculate final output commitment
+		let output_commit =
+			secp::add_excess(&swap.onion.commit, &peeled.0.excess).map_err(|e| Error {
+				message: e.to_string(),
+				code: ErrorCode::InternalError,
+				data: None,
+			})?;
+		let output_commit = secp::sub_value(&output_commit, fee).map_err(|e| Error {
+			message: e.to_string(),
+			code: ErrorCode::InternalError,
+			data: None,
+		})?;
+
+		// Verify the bullet proof and build the final output
+		if let Some(r) = peeled.0.rangeproof {
+			let secp = Secp256k1::with_caps(secp256k1zkp::ContextFlag::Commit);
+			secp.verify_bullet_proof(output_commit, r, None)
+				.map_err(|_| Error::invalid_params("RangeProof invalid"))?;
+		} else {
+			// milestone 3: only the last hop will have a rangeproof
+			return Err(Error::invalid_params("Rangeproof expected"));
+		}
+
+		let mut locked = SERVER_STATE.lock().unwrap();
+		if locked.contains_key(&swap.onion.commit) {
+			return Err(Error::invalid_params("swap already called for coin"));
+		}
+
+		locked.insert(
+			swap.onion.commit,
+			Submission {
+				excess: peeled.0.excess,
+				output_commit,

```

Comment:
`output_commit` already has `v*H + r*G + x_i*G - fee_i*H` from the `// Calculate final output commitment` code. Does a `Submission` contain a commitment that already had changes applied?
I'm guessing each mixnode will receive an `Onion` which they will peel to get `(Payload, Onion')`. This means each mixnode will get their own fee that they will check simulate picking it up by calling `secp::sub_value` the way you do it above. The resulting commitment will be compared with the next commitment that they received in the `Onion'.commitment` part after peeling, correct?

Reply:
That's correct. I pushed some changes that I hope make it easier to review and understand. The server code is no longer cluttered with json rpc boilerplate, so the tests should be easier to follow.